src/shogun/preprocessor/KernelPCA.cpp

```diff
@@ -80,43 +79,38 @@ bool CKernelPCA::init(CFeatures* features)
 		int32_t m = kernel_matrix.num_rows;
 		ASSERT(n==m)
 
-		float64_t* bias_tmp = SGMatrix<float64_t>::get_column_sum(kernel_matrix.matrix, n,n);
-		SGVector<float64_t>::scale_vector(-1.0/n, bias_tmp, n);
-		float64_t s = SGVector<float64_t>::sum(bias_tmp, n)/n;
-		SGVector<float64_t>::add_scalar(-s, bias_tmp, n);
+		SGVector<float64_t> bias_tmp = linalg::rowwise_sum(kernel_matrix);
+		linalg::scale(bias_tmp, bias_tmp, -1.0 / n);
+		float64_t s = linalg::sum(bias_tmp) / n;
+		linalg::add_scalar(bias_tmp, -s);
 
-		SGMatrix<float64_t>::center_matrix(kernel_matrix.matrix, n, m);
+		linalg::center_matrix(kernel_matrix);
 
-		float64_t* eigenvalues=SGMatrix<float64_t>::compute_eigenvectors(kernel_matrix.matrix, n, n);
+		SGVector<float64_t> eigenvalues(kernel_matrix.num_cols);
+		SGMatrix<float64_t> eigenvectors(
+		    kernel_matrix.num_rows, kernel_matrix.num_cols);
+		linalg::eigen_solver_symmetric(
+		    kernel_matrix, eigenvalues, eigenvectors);
 
+		m_transformation_matrix =
+		    SGMatrix<float64_t>(kernel_matrix.num_rows, kernel_matrix.num_cols);
+		linalg::zero(m_transformation_matrix);
+		auto args = CMath::argsort(eigenvalues);
 		for (int32_t i=0; i<n; i++)
 		{
 			//normalize and trap divide by zero and negative eigenvalues
-			for (int32_t j=0; j<n; j++)
-				kernel_matrix.matrix[i*n+j]/=CMath::sqrt(CMath::max(1e-16,eigenvalues[i]));
+			auto idx = n - args[i] - 1;
+			SGVector<float64_t> vec(
+			    eigenvectors.get_column_vector(idx), n, false);
+			linalg::add_col_vec(
```

Comment:
Yes, add it. Need to tell the SGVector to not free memory at the end though, and add a clear warning in the API doc that it is non-owning